src/core/connection.cpp

```diff
@@ -33,360 +33,631 @@
-    : state_(p_websocket ? State::AWAIT_CONNECTION
-                         : State::ERROR)
-    , p_websocket_(std::move(p_websocket))
-    , p_error_handler_(std::move(p_error_handler))
-{
-    assert(p_error_handler_);
-}
-
-bool Connection::login(const std::string& auth, Buffer* p_user_data)
-{
-    if (!p_websocket_)
-        return false;
+    void Connection::send_authentication_request()
+    {
+        assert(state_ == ConnectionState::AWAIT_AUTHENTICATION);
+        assert(p_auth_params_);
 
-    if (state_ != State::AWAIT_AUTHENTICATION)
-        throw std::logic_error("Cannot login() in current state");
+        MessageBuilder authentication_request(Topic::AUTH, Action::REQUEST);
+        authentication_request.add_argument(*p_auth_params_);
 
-    MessageBuilder areq(Topic::AUTH, Action::REQUEST);
-    if (!auth.empty())
-        areq.add_argument(auth);
-
-    if (send_(areq) != websockets::State::OPEN)
-        return false;
-
-    Buffer buffer;
-    parser::MessageList messages;
-
-    if (receive_(&buffer, &messages) != websockets::State::OPEN)
-        return false;
-
-    if (messages.size() != 1) {
-        close();
-        return false;
+        send(authentication_request);
     }
 
-    const Message& msg = messages.front();
-
-    if (msg.topic() == Topic::AUTH && msg.action() == Action::REQUEST && msg.is_ack()) {
-        assert(state_ == State::CONNECTED);
-
-        if (msg.num_arguments() == 0 && p_user_data)
-            p_user_data->clear();
-        else if (msg.num_arguments() == 1 && p_user_data)
-            *p_user_data = msg[0];
-
-        return true;
+    void Connection::close()
+    {
+        deliberate_close_ = true;
+        ws_handler_.close();
     }
 
-    if (msg.topic() == Topic::AUTH && msg.action() == Action::ERROR_INVALID_AUTH_DATA) {
-        assert(state_ == State::AWAIT_AUTHENTICATION);
-        p_error_handler_->onError(ErrorState::AUTHENTICATION_ERROR);
-        return false;
+    ConnectionState Connection::state()
+    {
+        return state_;
     }
 
-    if (msg.topic() == Topic::AUTH && msg.action() == Action::ERROR_INVALID_AUTH_MSG) {
-        p_error_handler_->onError(ErrorState::AUTHENTICATION_ERROR);
-        close();
-        return false;
+    void Connection::state(const ConnectionState state)
+    {
+        DEBUG_MSG("Connection state change: " << state);
+        state_ = state;
     }
 
-    if (msg.topic() == Topic::AUTH && msg.action() == Action::ERROR_TOO_MANY_AUTH_ATTEMPTS) {
-        p_error_handler_->onError(ErrorState::AUTHENTICATION_ERROR);
-        close();
-        return false;
-    }
+    void Connection::on_message(const Buffer &raw_message)
+    {
+        DEBUG_MSG(raw_message.size() << " bytes received");
+        Buffer buffer;
 
-    close();
-    return false;
-}
+        buffer.assign(raw_message.cbegin(), raw_message.cend());
+        buffer.push_back(0);
```

Comment:
it's a precondition for the parser (see https://github.com/deepstreamIO/deepstream.io-client-cpp/blob/finishing-split/src/core/parser.cpp#L77). Not sure why it doesn't add the trailing nulls itself though...